From 24705ccf48954c3df2fa8e6364d3d572626a35d2 Mon Sep 17 00:00:00 2001
From: Cosmin-Gabriel Samoila <cosmin.samoila@nxp.com>
Date: Tue, 23 Jan 2018 10:58:32 +0200
Subject: [PATCH 3238/5242] MLK-15033: ASoC: codecs: AK4458: Remove
 unsupported rates and channels

commit  044451aac769b3f716b85700dfdc0d1eeb56946d from
https://source.codeaurora.org/external/imx/linux-imx.git

Support only even number of channels greater than 2 and
rates multiple of 8000.

Signed-off-by: Cosmin-Gabriel Samoila <cosmin.samoila@nxp.com>
Reviewed-by: Shengjiu Wang <shengjiu.wang@nxp.com>
Reviewed-by: Daniel Baluta <daniel.baluta@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 sound/soc/fsl/imx-ak4458.c |   37 +++++++++++++++++++++++++++++++++++++
 1 file changed, 37 insertions(+)

diff --git a/sound/soc/fsl/imx-ak4458.c b/sound/soc/fsl/imx-ak4458.c
index 52d372e..cbaf31c 100644
--- a/sound/soc/fsl/imx-ak4458.c
+++ b/sound/soc/fsl/imx-ak4458.c
@@ -18,6 +18,7 @@
 #include <linux/i2c.h>
 #include <sound/soc.h>
 #include <sound/pcm_params.h>
+#include <sound/pcm.h>
 #include <sound/soc-dapm.h>
 
 struct imx_ak4458_data {
@@ -30,6 +31,15 @@ struct imx_ak4458_data {
 	SND_SOC_DAPM_LINE("Line Out", NULL),
 };
 
+static const u32 ak4458_rates[] = {
+	8000, 16000, 32000,
+	48000, 96000, 192000,
+};
+
+static const u32 ak4458_channels[] = {
+	1, 2, 4, 6, 8, 10, 12, 14, 16,
+};
+
 static int imx_aif_hw_params(struct snd_pcm_substream *substream,
 				struct snd_pcm_hw_params *params)
 {
@@ -72,8 +82,35 @@ static int imx_aif_hw_params(struct snd_pcm_substream *substream,
 	return ret;
 }
 
+static int imx_aif_startup(struct snd_pcm_substream *substream)
+{
+	struct snd_pcm_runtime *runtime = substream->runtime;
+	static struct snd_pcm_hw_constraint_list constraint_rates;
+	static struct snd_pcm_hw_constraint_list constraint_channels;
+	int ret;
+
+	constraint_rates.list = ak4458_rates;
+	constraint_rates.count = ARRAY_SIZE(ak4458_rates);
+
+	ret = snd_pcm_hw_constraint_list(runtime, 0, SNDRV_PCM_HW_PARAM_RATE,
+						&constraint_rates);
+	if (ret)
+		return ret;
+
+	constraint_channels.list = ak4458_channels;
+	constraint_channels.count = ARRAY_SIZE(ak4458_channels);
+
+	ret = snd_pcm_hw_constraint_list(runtime, 0, SNDRV_PCM_HW_PARAM_CHANNELS,
+						&constraint_channels);
+	if (ret)
+		return ret;
+
+	return 0;
+}
+
 static struct snd_soc_ops imx_aif_ops = {
 	.hw_params = imx_aif_hw_params,
+	.startup = imx_aif_startup,
 };
 
 static struct snd_soc_dai_link_component ak4458_codecs[] = {
-- 
1.7.9.5

